### PR TITLE
Update SCRUM model to new version

### LINKED ISSUE #7869: SCRUM upgrade to be more general
Work approved by Steering comittee and funded by AI.  See attached for project description
[Proposal for APSIM initiative SCRUM.docx](https://github.com/APSIMInitiative/ApsimX/files/11034756/Proposal.for.APSIM.initiative.SCRUM.docx)

### PR DESCRIPTION
**Pull request by HamishBrownPFR**
resolves #7869

This PR represents a major upgrade to the SCRUM model following development efforts partly funded by the APSIM initiative to make the model more generically relevant so it can be used to represent crops which currently lack detailed crop models in APSIM.

The upgrade to SCRUM includes:
- Major changes to the SCRUM model to allow for quick and simple parameterization of new crops
- Updating of documentation to reflect changes to the model
- Inclusion of a ScrumCropInstance model which provides a UI component for setting crop parameters
- Inclusion of a range of approaches for including SCRUM models in simulations via manager scripts
- Inclusion of a rotation model to simplify the job of making crop rotations using scrum models
- Inclusion of additional validation data and simulations for the SCRUM test set.
- Sensibility tests in the validation set
- A set of example simulations to demonstrate different ways SCRUM can be used
- A set of example ScrumCropInstances giving a start point for users to parameterize a wide range of crops.  These are based on a substantial effort to collate this data for these crops and use expert opinion to extrapolate values from similar crops where data could not be found.

### CODE REVIEW

**Comment by HamishBrownPFR** [comment]
There are still a few things here we are tidying up but getting the PR active to get it all running on the build system.  
@hol353 @hut104 @par456 @ric394 there is quite a bit in this.  If it is helpful, we can do a VC and I can run you through it all to help the review process.  This has already been OK'ed by reference pannel and we have addressed all the requirements raised there.  I also have a recording of the presentation to the Review panel which can go onto the website soon.  It covered SPRUM and STRUM as well so will wait until I have them all ready for release

**Comment by ric394** [comment]
@HamishBrownPFR Changes were made to the POStats collector yesterday afternoon, that's the error you're seeing. I'll investigate.

**Comment by ric394** [comment]
retest this please jenkins

**Comment by par456** [comment]
retest this please jenkins

**Comment by par456** [comment]
retest this please jenkins

File: Examples/Eucalyptus.apsimx [hunk 1]
```diff
@@ -23,7 +23,7 @@
         {
```
Comment (hol353) [inline Examples/Eucalyptus.apsimx:23]: Why is Eucalyptus.apsimx in this PR?

File: Examples/Rotation1.csv [hunk 2]
```diff
@@ -3,7 +3,7 @@ CropName,Potato,Wheat,OatForage,Carrot,TurnipForage
 EstablishDate,15-Oct,15-Apr,1-Mar,1-Sep,1-Mar
```
Comment (hol353) [inline Examples/Rotation1.csv:3]: Why has Rotation1.csv been moved to the Examples directory?

File: Models/PMF/SimplePlantModels/ScrumCropInstance.cs [hunk 3]
```diff
@@ -8,6 +8,8 @@
 using Models.PMF.Interfaces;
 using Models.PMF.Phen;
 using Models.Surface;
+using APSIM.Numerics;
+using DocumentFormat.OpenXml.Wordprocessing;
```
Comment (hol353) [inline Models/PMF/SimplePlantModels/ScrumCropInstance.cs]: I don't think you need ```using DocumentFormat.OpenXml.Wordprocessing;```

File: Models/PMF/SimplePlantModels/ScrumCropInstance.cs [hunk 4]
```diff
@@ -69,88 +75,181 @@ public class ScrumCropInstance : Model
...
             " Parameters defining growth pattern and biomass partition")]
-        [Description(" Harvest Index (0-1):")]
-        public double HarvestIndex { get; set; }
+        [Description(" Harvest Index (0.01-0.99):")]
+        public double HarvestIndex
+        {
+            get { return _HarvestIndex; }
+            set { _HarvestIndex = constrain(value, 0.01, 0.99); }
+        }
+        private double _HarvestIndex { get; set; }
```
Comment (hol353) [inline Models/PMF/SimplePlantModels/ScrumCropInstance.cs]: All privates to top of class

File: Tests/Validation/SoilTemperature/SoilTemperature.apsimx [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
```
Comment (hol353) [inline Tests/Validation/SoilTemperature/SoilTemperature.apsimx:1]: Remove SoilTemperature.apsimx?

File: Tests/Validation/Eucalyptus/Eucalyptus.apsimx [hunk 6]
```diff
@@ -1,13 +1,12 @@
 {
```
Comment (hol353) [inline Tests/Validation/Eucalyptus/Eucalyptus.apsimx:1]: Remove validation Eucalyptus.apsimx?

File: Tests/Validation/Eucalyptus/Eucalyptus.apsimx [hunk 7]
```diff
@@ -1,13 +1,12 @@
...
   "$type": "Models.Core.Simulations, Models",
-  "Version": 187,
```
Comment (hol353) [inline Tests/Validation/Eucalyptus/Eucalyptus.apsimx:3]: Why is SCRUM.json in UnitTests? Doesn't seem to be referenced anywhere.

**Review by hol353 — CHANGES_REQUESTED** [review]

File: Examples/Eucalyptus.apsimx (hunk 1 above)
Comment (HamishBrownPFR) [reply Examples/Eucalyptus.apsimx:23]: It needs to be.  There is a simulation that used SCRUM to represent a weed to test competition.  SCRUM is actually much better suited for this purpose now and it is a use I hadn't' thought about.

Comment (HamishBrownPFR) [reply Examples/Eucalyptus.apsimx:23]: Oh, Eucalyptus changed because I had to replace the old SCRUM with the updated version

File: Examples/Rotation1.csv (hunk 2 above)
Comment (HamishBrownPFR) [reply Examples/Rotation1.csv:3]: It needs to be there.  I have included a example with the SCRUM manager component which reads and writes its parameters to a csv file.  The intent was so rotations could be edited in the UI grid and via a grided csv editor (e.g Rons data edit)

File: Models/PMF/SimplePlantModels/ScrumCropInstance.cs (hunk 3 above)
Comment (HamishBrownPFR) [reply Models/PMF/SimplePlantModels/ScrumCropInstance.cs]: Removed

File: Models/PMF/SimplePlantModels/ScrumCropInstance.cs (hunk 4 above)
Comment (HamishBrownPFR) [reply Models/PMF/SimplePlantModels/ScrumCropInstance.cs]: I found it easier to have the private fields paired with the public fields they are working with.

File: Tests/Validation/SoilTemperature/SoilTemperature.apsimx (hunk 5 above)
Comment (HamishBrownPFR) [reply Tests/Validation/SoilTemperature/SoilTemperature.apsimx:1]: There is a test in soil temperature that uses SCRUM in the same way as the Eucalyptus test so this needs to be there

File: Tests/Validation/Eucalyptus/Eucalyptus.apsimx (hunk 6 above)
Comment (HamishBrownPFR) [reply Tests/Validation/Eucalyptus/Eucalyptus.apsimx:1]: There is a simulation in there that use SCRUM that had to be updated

File: Tests/Validation/Eucalyptus/Eucalyptus.apsimx (hunk 7 above)
Comment (HamishBrownPFR) [reply Tests/Validation/Eucalyptus/Eucalyptus.apsimx:3]: I don't understand this comment.  Can you please elaborate @hol353

File: Models/PMF/SimplePlantModels/ScrumCropInstance.cs (hunk 4 above)
Comment (HamishBrownPFR) [reply Models/PMF/SimplePlantModels/ScrumCropInstance.cs]: Moved to the top now

File: Tests/Validation/Eucalyptus/Eucalyptus.apsimx (hunk 7 above)
Comment (hol353) [reply Tests/Validation/Eucalyptus/Eucalyptus.apsimx:3]: image

Is this file necessary: Tests/UnitTests/Documentation/TestFiles/SCRUM.json?

File: Examples/Eucalyptus.apsimx (hunk 1 above)
Comment (hol353) [reply Examples/Eucalyptus.apsimx:23]: Given that there are a few .apsimx files in this PR and that SCRUM is in the release already and may have users, perhaps you better write a converter.

NOTE that my PR (#9940) will probably be merged in the next hour and it moves the converter.cs file into APSIM.Core. You will probably want to wait until that PR is merged.

Comment (HamishBrownPFR) [reply Examples/Eucalyptus.apsimx:23]: I don't think we can write a converter as we would need to pluck information out of managers and get in into the scrum management instance and information out of the deprecated version of SCRUM.json and put it into the a ScrumCropInstance.  At least it would not be a simple job and I don't think the effort would be justified given the few existing users and because it is not hard to update manually and that users should be aware that there has been a major version change with SCRUM (having to update manually will make them well aware).  There are notes to this effect in the doco

File: Tests/Validation/Eucalyptus/Eucalyptus.apsimx (hunk 7 above)
Comment (HamishBrownPFR) [reply Tests/Validation/Eucalyptus/Eucalyptus.apsimx:3]: Yes, that one is necessary because I have updated the doco to reflect the new version.   There is a unit test that checks if the doco has changed and the build will fail if you don't update it and commit it.

**Review by hol353 — APPROVED** [review]

**Comment by hol353** [comment]
Looking at the stats for the first time. The stats for Eucalyptus and SoilTemp are much worse. Why?

**Comment by HamishBrownPFR** [comment]
Hmmm.  There were not affected previously.  I will investigate

**Comment by LenLon** [comment]
Super curious to see what's new, great to see it finally happening :-)

**Comment by HamishBrownPFR** [comment]
@LenLon I think alot of people will find the new version of SCRUM really useful.  We are also adding a SPRUM (pasture) and STRUM (tree) model to the suite of simple uptake models that will go be available over the next month.  Are you using the current version of SCRUM currently.  I am asking because the old version will not work when we release the new version and it will not be possible to write a converter so users will have to change existing simulations to use the new version of SCRUM manually.  
Once the new models are in release we need to work out a good way to introduce people to the and train users on how to use them effectively.  Open to good ideas here.

**Comment by HamishBrownPFR** [comment]
@hol353 I have fixed the stats for Eucalyptus and Soil Temp to some extent and the diffs there are now small.  The problem was there was a script to harvest eucalyputs in both simulations that was catching the SCRUM harvest and responding to that.  The stats are not the same because the SCRUM model that was used to represent weeds has changed a bit.  There is not any weed data there to try to improve the fits for SCRUM any further.

**Comment by hol353** [comment]
@HamishBrownPFR Sorry to ask but can I get you to change some of your output variables. They are very long and can be shortened I think.

```
(Soil.OutputNutrientLayers.NH4(1)+Soil.OutputNutrientLayers.NO3(1)+Soil.OutputNutrientLayers.NH4(2)+Soil.OutputNutrientLayers.NO3(2))
```

can be shortened to 

```
Soil.OutputNutrientLayers.NH4(1:2)+Soil.OutputNutrientLayers.NO3(1:2)
```

Also

```
(sum(Soil.OutputNutrientLayers.NH4(1))+sum(Soil.OutputNutrientLayers.NO3(1)))
```
can be

```
Soil.OutputNutrientLayers.NH4(1)+Soil.OutputNutrientLayers.NO3(1)
```

This can apply to some of the other variables as well.

**Comment by HamishBrownPFR** [comment]
@hol353 not problem to ask this.  I would very much like to tidy these variable names up but need #9991 to be resolved first.  Unfortunately it is beyond my depth and I will need to prevail on one of the software team to help with that.
